### PR TITLE
Fix returning empty array in getEntityRecords for unknown entities

### PR DESCRIPTION
**Pull request by kevin940726**
## Description

`getEntityRecords` currently will return an empty array if the entity is unknown (not fetched), but it will also later return `null` when the entity is loaded (via `getQueriedItems`), until finally return the list of records when the data is fully fetched.

This behavior doesn't make sense to me, I think we should also return `null` if the entity is unknown, to align with the `null` value when the data is still fetching. By returning `null` for the initial request, the end developers can also better distinguish between a loaded empty list and a loading state.

There are some former discussion about this: https://github.com/WordPress/gutenberg/pull/19498#discussion_r462630069, https://github.com/WordPress/gutenberg/pull/21630#issuecomment-614881500. I believe the consensus was to always return `null`.

This shouldn't be a breaking change since it already returns `null` in other cases, and we could treat the original behavior as a bug.

## How has this been tested?

Unit tests should pass.

## Types of changes

Bug fix

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/core-data/src/test/selectors.js
Comment (kevin940726) [inline packages/core-data/src/test/selectors.js:163]: This doesn't make sense to me either. Why would we want to return `true` for `hasEntityRecords` when the entity is still unknown?

**Comment by talldan** [comment]
@kevin940726 It'd be good to detail what led you to making this fix. Does it fix a known bug or is it an observation you made about some code that could be improved?

File: packages/core-data/src/test/selectors.js
Comment (talldan) [reply packages/core-data/src/test/selectors.js:163]: Yeah, agree, that doesn't seem right given the documentation for `hasEntityRecords`:
> Returns true if records have been received for the given set of parameters, or false otherwise.

The key word being 'received'. That queried data prepopulates for known entity configurations is irrelevant, those records weren't received in the typical way we use that term.

I also find the name `hasEntityRecords` confusing since it should really be called `hasReceivedEntityRecords`. The current name would lead me to believe an empty array would return `false` too, but the function definition instead just checks `Array.isArray`.

And with the way it works in `trunk` it should perhaps just be called `hasEntityConfiguration` 😄

**Comment by kevin940726** [comment]
> @kevin940726 It'd be good to detail what led you to making this fix. Does it fix a known bug or is it an observation you made about some code that could be improved?

Sure! I found this bug while I was working on the templates list page. Specifically, I was calling `getEntityRecords('postType', 'wp_template')` in a `useSelect` trying to get a list of all the templates.

I put a `console.log` in the same render function, and discovered that the initial result is always `[]`, while the next few renders logged `null` until it received the data.

Since that the initial data is always `[]`, there's no trivial way to determine if the current result is "not yet started loading" or "loaded but no results". Hence the motivation of this PR.

**Comment by mcsf** [comment]
This makes sense, and semantically I prefer it, but I'm not sure of what unexpected consequences it could have. So I'd defer to @youknowriad. (Also cc @ntsekouras — no action needed, just thought you might be interested in the topic.)

**Comment by noisysocks** [comment]
> I'm not sure of what unexpected consequences it could have

Neither. I'd say let's document the change in `CHANGELOG.md` and ensure that there's plenty of time for testing (i.e. don't land this in WP 5.9).

**Review by youknowriad — APPROVED** [review]
I also agree with this semantically speaking but I know we had a few surprises in the past in this kind of changes so let's be vigilant about any subtle breakage.

**Review by adamziel — APPROVED** [review]
Ditto what Riad and Miguel said

**Pull request merged**
